Destiny.ScrimTracker.Logic/Services/MatchMakingService.cs

```diff
@@ -0,0 +1,254 @@
+                    Guardian = guardian,
+                    GuardianEfficiency = guardianEff,
+                    GuardianElo = guardianElo
+                };
+                
+                guardians.Add(snapshot);
+            }
+            
+            var teams = RandomizeTeams(guardians, teamSize);
+
+            teams = Anneal(teams);
+            
+            return teams;
+        }
+
+        private IEnumerable<MatchMadeTeam> RandomizeTeams(IEnumerable<GuardianSnapshot> guardians, int teamSize)
+        {
+            var numberOfTeams = guardians.Count() / teamSize;
+            var guardianArray = ShuffleGuardiansList(guardians);
+            
+            var teams = new List<MatchMadeTeam>();
+            for (var i = 0; i < numberOfTeams; i++)
+            {
+                var team = new MatchMadeTeam();
+                var teamGuardians = new List<GuardianSnapshot>();
+                var teamElo = 0.0;
+                var teamEfficiency = 0.0;
+                
+                for (var j = i * teamSize; j < i * teamSize + teamSize; j++)
+                {
+                    teamGuardians.Add(guardianArray[j]);
+                    teamElo += guardianArray[j].GuardianElo;
+                    teamEfficiency += guardianArray[j].GuardianEfficiency;
+                }
+
+                team.Guardians = teamGuardians;
+                team.TeamElo = teamElo / teamSize;
+                team.TeamEfficiency = teamEfficiency / teamSize;
+                
+                teams.Add(team);
+            }
+
+            return teams;
+        }
+
+        private IEnumerable<MatchMadeTeam> Anneal(IEnumerable<MatchMadeTeam> teams)
+        {
+            var temperature = 400.0;
+
+            var iterations = 0;
+            while (temperature > _epsilon)
+            {
+                iterations++;
+
+                var (teamIndex1, teamIndex2) = GetRandomTeamIndexesForSwap(teams);
+                var nextTeamConfiguration = SwapTwoPlayers(teams, teamIndex1, teamIndex2);
+                
+                var teamElos = GetTeamElos(teams).ToArray();
+                var newTeamElos = GetTeamElos(nextTeamConfiguration).ToArray();
+                
+                var startingDelta = CalculateDelta(teamElos[teamIndex1], teamElos[teamIndex2]);
+                var newConfigurationDelta = CalculateDelta(newTeamElos[teamIndex1], newTeamElos[teamIndex2]);
+
+                var delta = newConfigurationDelta - startingDelta;
+                if (delta < 0)
+                {
+                    teams = KeepNextConfiguration(nextTeamConfiguration);
+                }
+//                else
+//                {
+//                    var probabilityToKeepChange = _random.NextDouble();
+//
+//                    if (probabilityToKeepChange < Math.Exp(-delta / temperature))
+//                    {
+//                        teams = KeepNextConfiguration(nextTeamConfiguration);
+//                    }
+//                }
```

Comment:
I'm trying to reason through the probability function you're got here and the implications it has....

`Math.Exp(x)` is `e^x`, which approaches zero for negative values of `x`, is 1 when `x` is 0 and increases exponentially when `x` is positive.

If `x` is positive, then `delta` is negative, which means its less than `0` which means we won't hit that case.
If `x` is zero, then `Math.Exp` is 1, so we will always keep that case. This is interesting and probably not what we want, but also unlikely.

That leaves cases where `delta` is positive, which will make `x` negative. The probability will be close to 1 when `x` is barely negative, for example when `delta` is very small. When `0 > x > -1`, (ie when `delta` is smaller than `temperature`), the probability will be extremely close to 1. When `delta` is large, the probability will be smaller. This means we favor small `delta` changes, and our tolerance for what is small gets smaller as the temperature decreases.

_If_ most changes result in _worse_ configurations (meaning a positive `delta`), _and_ most configurations are relatively similar to one another (because ELOs are similar), then it would seem that this probability will almost always result in us keeping the "worse" configuration.  If by chance, the algorithm comes up with a better solution, we'll use that until a worse one is encountered, and then we'll use the worse one. In essence we'll likely walk away from a good solution until we hit some point where a random change is just as likely to be bad as it is good, then we'll oscillate between moving towards a good solution and a bad solution. We'll do this until `temperature` is very small, then we'll only start moving towards good solutions. The thing is that by the time `temperature` is small enough, we don't have very many iterations left in the algorithm, we're too "cool" essentially to walk our way back to a global maximum. 

That's my theory anyway. You may just do something like `Math.Exp(-1000 * delta / temperature)`, or start `temperature = 4` and `_alpha = .00999`.

Reply:
Yea.. I just need to sit down and dig into the math and modeling here, I think. I also think you're thinking about this in an inverted kinda way. Delta is guaranteed to be > 0 in this case. If it's not, then we don't even hit the e^x probability function.

Reply:
> That leaves cases where `delta` is positive, which will make `x` negative

That's the case I talk about the most. It could also be `0`, which I think is unlikely, and would be weird.

Reply:
oh right i can read.. whoooops 😂 i did try the suggestions you left at the bottom there but they were so small they were rounded to 0 lol and yea, i don't think we need to start with temp = 400, just overkill but i dunno

Reply:
Maybe then leave temp and alpha alone and try multiplying Delta by 1000.

You could do some data gathering by logging out all the values used on every iteration along with a) did we use the probability branch and b) on that branch did we actually take the step backward.

Non-deterministic functions are tough to debug, but logging out lots of information can help spot weirdness.